lib-multisrc/gigaviewer/src/eu/kanade/tachiyomi/multisrc/gigaviewer/GigaViewer.kt

```diff
@@ -180,6 +181,73 @@ abstract class GigaViewer(
         return chapters
     }
 
+    protected fun paginatedChaptersRequest(referer: String, aggregateId: String, offset: Int): Response {
+        val headers = headers.newBuilder()
+            .set("Referer", referer)
+            .build()
+
+        val apiUrl = baseUrl.toHttpUrl().newBuilder()
+            .addPathSegment("api")
+            .addPathSegment("viewer")
+            .addPathSegment("pagination_readable_products")
+            .addQueryParameter("type", "episode")
+            .addQueryParameter("aggregate_id", aggregateId)
+            .addQueryParameter("sort_order", "desc")
+            .addQueryParameter("offset", offset.toString())
+            .build()
+            .toString()
+
+        val request = GET(apiUrl, headers)
+        return client.newCall(request).execute()
+    }
+
+    protected fun chapterListParsePaginated(response: Response): List<SChapter> {
+        val document = response.asJsoup()
+        val referer = response.request.url.toString()
+        val aggregateId = document.selectFirst("script.js-valve")!!.attr("data-giga_series")
+
+        val chapters = mutableListOf<SChapter>()
+
+        var offset = 0
+
+        // make initial request
+        var result = paginatedChaptersRequest(referer, aggregateId, offset)
+        var resultData = json.decodeFromString<GigaViewerEpisodesDto>("{\"episodes\": ${result.body.string()}}")
+
+        // repeat until the offset is too large to return any chapters, resulting in an empty list
+        while (resultData.episodes.isNotEmpty()) {
+            resultData.episodes.mapTo(chapters) { element ->
+                SChapter.create().apply {
+                    name = element.title
+                    if (chapterListMode == CHAPTER_LIST_PAID && element.status?.label != IS_FREE) {
+                        name = YEN_BANKNOTE + name
+                    } else if (chapterListMode == CHAPTER_LIST_LOCKED && element.status?.label == UNPUBLISHED) {
+                        name = LOCK + name
+                    }
+                    date_upload = element.display_open_at?.toDate() ?: 0L
+                    scanlator = publisher
+                    setUrlWithoutDomain("/episode/${element.readable_product_id}")
+                }
+            }
+            // increase offset and make next request
+            offset += resultData.episodes.size
+            result = paginatedChaptersRequest(referer, aggregateId, offset)
+            resultData = json.decodeFromString<GigaViewerEpisodesDto>("{\"episodes\": ${result.body.string()}}")
```

Comment:
```suggestion
            resultData = result.parseAs<List<GigaViewerPaginationReadableProduct>>()
```

Doesn't this work? Why wrap it by string concatenation?